File: Java/src/CourseScheduleII.java
import java.util.ArrayList;
import java.util.Stack;

public class CourseScheduleII {
	private boolean [] visited;
	private boolean [] onStack;
	private ArrayList<ArrayList<Integer>> graph;
	private Stack<Integer> stack;
	public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] result = new int [numCourses];
        
        graph = new ArrayList<ArrayList<Integer>>();
        stack = new Stack<Integer>();
        for(int i = 0; i < numCourses; i++) {
        	graph.add(new ArrayList<Integer>());
        }
        
        for(int i = 0; i < prerequisites.length; i++) {
        	graph.get(prerequisites[i][1]).add(prerequisites[i][0]);
        }
        visited = new boolean [numCourses];
        onStack = new boolean [numCourses];
        
        
        for(int i = 0; i < numCourses; i++) {
        	if(dfs(i)) return new int [0];
        }
        
        int index = 0;
        while(!stack.isEmpty()) {
        	result[index++] = stack.pop();
        }
        
        
        
        return result;
    }
	
	//return if contain cycle
	private boolean dfs(int node) {
		if(onStack[node]) return true;
		if(visited[node]) return false;
		
		onStack[node] = true;
		visited[node] = true;
		
		for(int childNode : graph.get(node)) {
			if(dfs(childNode)) return true;
		}
		
		onStack[node] = false;
		
		stack.push(node);
		return false;
	}
}

File: Python/CourseScheduleII.py
from typing import List

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.graph = {}
        for l in prerequisites:
            if l[0] not in self.graph:
                self.graph[l[0]] = []
            self.graph[l[0]].append(l[1])

        self.visited = [False] * numCourses
        self.onStack = [False] * numCourses
        self.res = []

        for i in range(numCourses):
            if self.dfs(i):
                return []
        
        # self.res.reverse()
        return self.res

    def dfs(self, n):
        if self.onStack[n]:
            return True
        if self.visited[n]:
            return False
        self.onStack[n] = True
        self.visited[n] = True

        if n in self.graph:
            for adj in self.graph[n]:
                if self.dfs(adj):
                    return True
        
        self.onStack[n] = False
        self.res.append(n)
        return False

test = Solution()
prerequisites = [[1,0],[2,0],[3,1],[3,2]]
res = test.findOrder(4, prerequisites)
print(res)